Add `SparseMatrix_JDS` class

A-CGray opened the pull request:
# Purpose

This PR adds a class that inherits from `SparseMatrix` and implements the Jagged Diagonal (JDS) storage format

Closes #22 
Closes #4 

# Definition of Done

### The following methods are implemented:
- [ ] Constructor
- [ ] Destructor 
- [ ] `assembleStorage`
- [ ] `matVec`
- [ ] `unassemble`
- [ ] `getFormat`

### The following methods are documented (e.g have docstrings):
- [ ] Constructor
- [ ] Destructor 
- [ ] `assembleStorage`
- [ ] `matVec`
- [ ] `unassemble`
- [ ] `getFormat`

### The code is formatted correctly:
- [ ] Private attribute names start with `_`
- [ ] Attributes are always accessed via `this->`
- [ ] Method names use `camelCase`

### The code is correct:
- [ ] All unit tests are passing
- [ ] The docs build correctly

A-CGray commented:
@bkochuna we still have nobody assigned to write the `unassemble` method for this class, are there still people with duplicate issues we can assign that to?

bkochuna commented:
I updated the #61 testing infrastructure branch to fix a few things with CMake. If you merge this branch in now (and manually resolve the merge conflicts) it'll work.

I'm also concerned with what changed in the COO matrix type that the example in `/examples` fails to build now.

A-CGray commented:
> I updated the #61 testing infrastructure branch to fix a few things with CMake. If you merge this branch in now (and manually resolve the merge conflicts) it'll work.

Great thanks, it does work now, I find it weird that CMake doesn't figure out that running the tests depends on compiling them, I was having issues that turned out to be caused by forgetting to run `make` before `make test`

> 
> I'm also concerned with what changed in the COO matrix type that the example in `/examples` fails to build now.

Nothing changed in the COO class, but I added `_unAssemble` and `computeMatVecProduct` as pure virtual functions in the base class which made the COO abstract and thus the example wouldn't compile, I've changed it now to use the JDS class.

A-CGray commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -0,0 +1,237 @@
+    cout << "Hello from SparseMatrix_JDS Destructor!" << endl;
+
+    if (this->_colIndices != NULL)
+      free(this->_colIndices);
```
@jasminylim I think you need to use `delete[]` here instead of `free` because I created these arrays using `new` rather than `malloc`.

jasminylim replied:
Issue should be fixed

A-CGray commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -228,6 +228,21 @@ namespace SpMV {
+    int nonzero;
+    //
+    for (i = 0; i < this->_maxRowSize; i++) {
+      nonzero = this->_jdPtrs[i + 1];
```
@umichyujl  There is a method called `getNumNonZeros` you can call to get the number of nonzeros in the matrix

A-CGray commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -228,6 +228,21 @@ namespace SpMV {
+    for (i = 0; i < this->_maxRowSize; i++) {
+      nonzero = this->_jdPtrs[i + 1];
+    }
+    // number of nonzero rows
+    nnrows = nonzero;
+    // number of nonzero columns
+    nncols = nonzero;
```
@umichyujl  Number of rows = number of columns = number of nonzeros in the matrix?

umichyujl replied:
Yeah... This is what I can think of...

umichyujl replied:
I will ask professor today for the function of getFormat. I am really confused.

A-CGray commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -228,6 +228,21 @@ namespace SpMV {
+    // number of nonzero columns
+    nncols = nonzero;
+    //
+    return SparseMatrix(nnrows, nncols);
```
@umichyujl This returns an empty nnrows x nncols Matrix, don't we also need to set all the nonzero entries from the current matrix in this new one too?

umichyujl replied:
I think setting the nonzero entries is done by setCoefficient based on the sample main function? I do not know if I need to set these values.

umichyujl commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -230,19 +230,24 @@ namespace SpMV {
-    nnrows = nonzero;
-    // number of nonzero columns
-    nncols = nonzero;
+    SparseMatrix<fp_type> B;
```
@A-CGray  I am pretty sure that I need to return a SparseMatrix type. Could you please double-check the testing files? Thanks!

A-CGray replied:
My best guess is that `getFormat` is supposed to return the matrix in a different format, in that case it should take in an input (a string or an enum I guess) that tells you what format to return, and it should return a pointer to a `SparseMatrix` (`SparseMatrix<fp_type>*`)

umichyujl replied:
Returning a pointer could be reasonable. If we use SparseMatrix<fp_type>* here for the return type, how can I return the matrix in the input format? I will change to the pointer first.

A-CGray replied:
A `SparseMatrix<fp_type>*` is allowed to point to any class that inherits from the `SparseMatrix` class

A-CGray replied:
Hey @umichyujl , I don't think the line `SparseMatrix<fp_type> B;` will work for a few reasons:

1) SparseMatrix is an abstract class so you cannot create an instance of it like this
2) Even if SparseMatrix wasn't an abstract class, declaring the variable like this means that B will be deleted from memory at the end of this function (any variable created inside a set of { } gets deleted when the program leaves the { })

I think you need to use the `new` command to create a matrix that will stay in memory after the function ends, something like this:

```cpp
template <class fp_type>
  SparseMatrix<fp_type> *SparseMatrix_JDS<fp_type>::getFormat(string fmt) {

    bool needToReassemble = false;
    if (this->_state == assembled) {
      _unAssemble();
      needToReassemble = true;
    }

    // Create pointer to new matrix that will be returned
    SparseMatrix<fp_type> *ptr_A = nullptr;

    // --- Create the new matrix in the requested format for ptr_a to point to ---
    if (fmt == "DEN") {
      ptr_A = new SparseMatrix_DEN<fp_type>(this->_nrows, this->_ncols);
    }
    else if (fmt == "COO") {
      ptr_A = new SparseMatrix_COO<fp_type>(this->_nrows, this->_ncols);
    }
    else if (fmt == "CSR") {
      ptr_A = new SparseMatrix_CSR<fp_type>(this->_nrows, this->_ncols);
    }
    else if (fmt == "JDS") {
      ptr_A = new SparseMatrix_JDS<fp_type>(this->_nrows, this->_ncols);
    }
    else if (fmt == "ELL") {
      ptr_A = new SparseMatrix_ELL<fp_type>(this->_nrows, this->_ncols);
    }

    // Copy the nonzero entry data to the new matrix and assemble it
    ptr_A->_buildCoeff = this->_buildCoeff;
    ptr_A->assembleStorage();

    // Reassemble this matrix if needed
    if (needToReassemble) {
      this->assembleStorage();
    }

    return ptr_A;
  }
  ```

umichyujl replied:
Thanks a lot Alasdair! Do we exactly need to reassemble the matrix? I talked with professor before and it seems that we only need to unassemble the original one and assemble the new matrix. What is the function of needToReassemble here? Is it relevant to the unit testing part? Or we need the assembled one to do some other things?

A-CGray replied:
I just thought it would be good practice to put the original matrix back into it's assembled form if that's how it started out, but if Prof Kochunas said otherwise then I'd go with what he said

umichyujl replied:
Okay, I will rewrite the code. Thanks again for your kind help! 👍 💯

ShiboTan commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -0,0 +1,343 @@
+    // Create pointer to new matrix that will be returned
+    SparseMatrix<fp_type> *ptr_A = nullptr;
+
+    // --- Create the new matrix in the requested format for ptr_a to point to ---
```
please delete ---

ShiboTan replied:
@umichyujl

ArshSaja commented:
It seems like JDS is good to go unless you are currently fixing anything. The tests are failing because the building script doesn't have everything right now in the current brnach (I think when all PRs are merged, it should work).

A-CGray commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -0,0 +1,343 @@
+                                              size_t jd_ptr[],
+                                              size_t max_row_size)
+      : SparseMatrix<fp_type>::SparseMatrix(nrows, ncols) {
+    this->_nnz = (size_t)sizeof(jdiag) / sizeof(jdiag[0]);
```
Hey, @tcoonsUM , this line is giving me a bunch of compiler warnings:

```
/home/ali/repos/ners570f22-SpMV/src/SparseMatrix_JDS.hpp:117:33: warning: ‘sizeof’ on array function parameter ‘jdiag’ will return size of ‘float*’ [-Wsizeof-array-argument]
  117 |     this->_nnz = (size_t)sizeof(jdiag) / sizeof(jdiag[0]);
```

I think since we now have max_row_size available it would be better just to use `this->_nnz = jd_ptr[max_row_size];`

tcoonsUM replied:
Pointers are weird. Fixed via https://github.com/bkochuna/ners570f22-SpMV/pull/34/commits/eda2a37512fe508670c4db08a3f4fb8402ac9b7d

A-CGray replied:
amen

A-CGray commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -0,0 +1,343 @@
+    this->_nnz = (size_t)sizeof(jdiag) / sizeof(jdiag[0]);
+    vector<size_t> vecPerm(perm, perm + nrows);
+
+    this->_state = building;
+    this->_rowPerm = vecPerm;
+    this->_values = jdiag;
+    this->_colIndices = col_ind;
+    this->_jdPtrs = jd_ptr;
+    this->_maxRowSize = max_row_size;
+    this->_state = assembled;
```
@tcoonsUM , I am getting some errors when running the unit tests that I think are coming from here. If you look at [this code](https://github.com/bkochuna/ners570f22-SpMV/blob/2e50c8b46f826bedc5964890c78566a06c2faca6/src/SparseMatrix_JDS.hpp#L96-L102) you can see that `this->_colIndices`, `this->_jdPtrs` and `this->_values` are not arrays but actually pointers (this required so that they can be used to represent arrays of different sizes)

This means you can't just run `this->_colIndices = col_ind;`, because the JDS destructor is going to try to delete whatever `this_colIndices` is pointing at, and `col_ind` might be a piece of memory that can't be deleted.

I think the solution is to allocate some memory for `this->_colIndices` exactly like is done [here](
https://github.com/bkochuna/ners570f22-SpMV/blob/2e50c8b46f826bedc5964890c78566a06c2faca6/src/SparseMatrix_JDS.hpp#L195-L197) and then copy the entries of `col_ind` into `this->_colIndices`, same goes for `this->_jdPtrs` and `this->_values`

tcoonsUM replied:
Ah that is a good point. To copy over the entries of each array/vector to our class variables, are you envisioning or a for loop or is there a nicer copy() function that you had in mind?

A-CGray replied:
Not that I know of, just gotta loop over it

tcoonsUM replied:
Alright this seemed to do it on my end: https://github.com/bkochuna/ners570f22-SpMV/pull/34/commits/282a588679165b220ce4bd763c8628035e6af84f

umichyujl commented on src/SparseMatrix_JDS.hpp:
```diff
@@ -0,0 +1,343 @@
+    SparseMatrix<fp_type> *ptr_A = nullptr;
+
+    // --- Create the new matrix in the requested format for ptr_a to point to ---
+    // if (fmt == "DEN") {
```
Hi @A-CGray, I think these "comment" lines are for compiling? The code itself is right, but we are not yet merging all the .hpp files together so we can not call CSR/ELL here. Correct me if I am wrong, thanks!

A-CGray replied:
Yes, you're right, I think we may have to merge this PR and some of the other matrix classes first, and then we can uncomment this code in a separate PR once the other matrix types are available

umichyujl replied:
Great, thanks!

umichyujl commented:
> 

Thanks for doing this Alasdair! I've commented out the testing lines for `getFormat` in the test folder.

A-CGray commented:
@ArshSaja @ShiboTan this is ready to go!

ArshSaja commented:
Nicely Done! @A-CGray

A-CGray commented:
@ArshSaja can you merge this?
